docs/decisions/0006-handling-learner-initiated-unenrollments.rst

```diff
@@ -0,0 +1,100 @@
+0006 handling learner initiated unenrollments
+#############################################
+
+Status
+******
+
+**Draft** May 2023
+
+Context
+*******
+
+The enterprise-subsidy service needs to maintain records of transaction reversals for any associated platform
+fulfillment cancellation. These reversals represent credit, up to the amount allocated by the original transaction,
+returned to the learner's customer's ledger. As there are many sources of a learner initiated unenrollments/
+cancellations, the enterprise subsidy service must take on the responsibility of monitoring the state of enterprise
+fulfillments. This is necessary to ensure that any learner initiated unenrollment or cancellation of an enterprise 
+fulfillment object will be evaluated and that corresponding reversals are written if/when applicable. 
+
+Decision
+********
+
+In a pre-event bus driven architectural world, the enterprise subsidy service will monitor the state of recently 
+modified, unenrolled enterprise fulfillment objects. It will do so by running a management command which
+will hit the platform enterprise fulfillment `API 
+<https://github.com/openedx/edx-enterprise/blob/master/enterprise/api/v1/views.py#L576>`_ somewhere between once and
+twice a day. The view will return a list of enterprise fulfillment objects which have been cancelled after the provided
+modified at timestamp filter. Notably, we provide a modified at time window that is smaller than the frequency at which
+the job runs. With these records, the enterprise subsidy service will 
+- First check for the existence of a transaction reversal associated with the fulfillment object's transaction and 
+  idempotency key. This is to prevent multiple reversal to be written for the same cancellation action. A learner
+  initiated unenrollment reversal's idempotency key will be formatted as 
+  `unenrollment-reversal-<enterprise_fulfillment_id>-<enterprise_enrollment_modified>`.
```

Comment:
This key won't fly in the case of a learner that redeems, unenrolls, redeems again, and unenrolls again, since the enterprise fulfillment UUID and enterprise enrollment ID with both be the same for both unenrollment attempts.  You probably have to incorporate the transaction UUID too.  But then if a reversal enters "failed" state, re-attempts to unenroll will produce idempotency_key collisions again, so you probably also have to _version_ the idempotency_key for every additional historical failed reversal.

Reply:
I don't necessarily think we should block on this since I'm not sure it's a v1 thing, but might be worth calling out in the ADR.

Reply:
Ah, didn't realize `<enterprise_enrollment_modified>` was a date.  That's probably fine then!